libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -99,6 +103,73 @@ string ABIFunctions::tupleEncoder(
 	});
 }
 
+string ABIFunctions::tupleDecoder(TypePointers const& _types, bool _fromMemory)
+{
+	string functionName = string("abi_decode_tuple_");
+	for (auto const& t: _types)
+		functionName += t->identifier();
+	if (_fromMemory)
+		functionName += "_fromMemory";
+
+	return createFunction(functionName, [&]() {
+		solAssert(!_types.empty(), "");
+
+		TypePointers decodingTypes;
+		for (auto const& t: _types)
+			decodingTypes.emplace_back(t->decodingType());
+
+		Whiskers templ(R"(
+			function <functionName>(headStart, dataEnd) -> <valueReturnParams> {
+				switch slt(sub(dataEnd, headStart), <minimumSize>) case 1 { revert(0, 0) }
```

Comment:
Do we have any kind of assertions that `headStart < dataEnd`? Otherwise this could overflow.

Reply:
Note that this is `slt`, i.e. signed comparison, so this case is caught here.